---
title: "EdgeR_orf"
author: "Dennis Amnebrink"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
      toc: yes
      toc_float:
        collapse: no
      fig_caption: yes
      code_folding: hide
      number_sections: true
---

# EdgeR-analysis 
EdgeR-analysis of contrast TNvsC for day 17. The analysis is carried out on orf-level, allowing for identification of taxa specific responses. As compared to KO which look at the total regulation of a function irrespective of smaller taxa variations.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r constants, message=FALSE, warning=FALSE}
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(data.table)
library(edgeR)
library(kableExtra)
library(viridis)
```

```{r file-loading, message=FALSE, warning=FALSE, cache=TRUE}
#File to connect NGI ID with our own classification ID, add the Âµ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz")

#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm)

rm_values <- read_tsv("../data/Rm_values.tsv")

target_genes <- read_tsv("../data/list_of_genes.tsv")
```

```{r prok-filter, message=FALSE, warning=FALSE, cache=TRUE}
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```
# EdgeR- analysis
```{r, edgeR_day17, message=FALSE, warning=FALSE, results="hide", cache=TRUE}
 bbmap_mat <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  group_by(sample, orf) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  select(sample, orf, count) %>%
  spread(sample, count, fill = 0) %>%
  column_to_rownames("orf") %>%
  as.matrix()

y <- DGEList(bbmap_mat)

#Changing the grouping to the known treatments from mesocosm.
y$samples$group = sample_ID %>% filter(day == "17") %>% select(sample, treatment) %>%  pull(treatment)

#It has been correctly done as they match with the original sample file
groupF <- y$samples$group

#Another way of verifying
match(rownames(y$samples), colnames(bbmap_mat))

y <- DGEList(counts=bbmap_mat, group=groupF)

keep <- filterByExpr(y, group=groupF)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y$samples

#The design-matrix doesn't like non-numeric categories, so here is a numeric corresponding to orignal levels, key =
# 1 = T, 2 = C, 3 = N, 4 = TN
design_m <- y$samples %>% select(group) %>%
  mutate(group = c(1,1,1,2,2,2,3,3,3,4,4,4))

y <- estimateDisp(y, design = design_m)

#plotBCV(y)

#Creating a designmatrix (this looks like the correct structure), C is the base treatment, exlcuding intercept column.
trial_mat <- model.matrix(~0+groupF)

fit <- glmQLFit(y, trial_mat)

#Comparing C vs T
colnames(trial_mat) = c("C","N","TN","t")

fit <- glmQLFit(y, trial_mat)

TNvsC <- makeContrasts(TN - C, levels=trial_mat)
```

```{r, joining_contrasting_days, message=FALSE, warning=FALSE, results="hide", cache=TRUE}
qlf <- glmQLFTest(fit, contrast=TNvsC)

day17 <- topTags(qlf, n = NULL) %>%
  as.data.frame() %>%
  mutate(day = "17") %>%
  rownames_to_column("orf") %>%
  select(-FDR)

edgeR_orfs <- day17 %>%
  mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down"))

write_tsv(edgeR_orfs, "../results/edgeR_orfs.tsv")
```

After EdgeR-analysis the resulting genes produced the following output also expressed as the table below.   
Out of a total of `r length(edgeR_orfs$orf)` orfs, `r length(edgeR_orfs %>% filter(type == "Up") %>% pull(orf))` are upregulated in the TN-treatment. While `r length(edgeR_orfs %>% filter(type == "Down") %>% pull(orf))` are upregulated in the C treatment.
```{r,  message=F, warning=FALSE}
edgeR_orfs <- read_tsv("../results/edgeR_orfs.tsv")

kable(edgeR_orfs %>%
  group_by(type) %>%
  tally() %>%
  ungroup() %>%
  rename("No. of genes" = "n")) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

```{r edgeR_plot, message=F, warning=FALSE, fig.cap='Gene abundance against logFC, Up denotes upregulated in the TN treatment, down indicates upregulated in the C-treatment. Horizontal dashed green lines indicates DIFF_TRESHOLD logFC treshold.'}
#Setting colours
colours <- c("#ff0000","#000000","#4248f5")

edgeR_orfs %>%
  ggplot(mapping = aes(x = logFC, y = logCPM)) +
  geom_point(aes(colour = type, size = type)) +
  scale_color_manual("Type",values = colours) +
  scale_size_manual(values = c("Up" = 1.5, "non-DE" = 0.2, "Down" = 1.5), guide = "none") +
  geom_vline(xintercept = DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  geom_vline(xintercept = -DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  theme_classic() +
  coord_flip() +
  facet_wrap(~ day)
```

## Target genes in the DE-genes
Here the general differential expression is presented, then focused on the presence of target genes among the up and down-regulated genes. It is a general response of ATPases seen in the class of Betaproteobacteria.The cell shape gene [ccmA](https://pubmed.ncbi.nlm.nih.gov/20510929/) is also shown to be upregulated in Deltaproteobacteria. The tables below show all the target genesthar are differentially expressed in the TN and C treatment respectively.
```{r, message=F, warning=FALSE}
#Investigating if orfs are in the list of target genes.
target_genes_present <- edgeR_orfs %>% #83604 rows 
  inner_join(eggnogs %>% distinct(orf, Preferred_name), by = "orf") %>% #Loosing unannotated orfs here #81 698 rows
  inner_join(target_genes %>% rename(Preferred_name = "gene"), by = "Preferred_name")  # 1831 rows 
  
list_overview <- target_genes_present %>%
  group_by(type) %>%
  tally() %>%
  ungroup()
```
Among the orfs annotated in the target gene list,  `r list_overview %>% filter(type == "Up") %>% pull(n)` were upregulated,   `r list_overview %>% filter(type == "Down") %>% pull(n)` were downregulated and lastly `r list_overview %>% filter(type == "non-DE") %>% pull(n)` were not differentially expressed.

```{r, message=F, warning=FALSE}
kable(target_genes_present %>%
  filter(type == "Up") %>%
  select(-Category) %>%
    inner_join(taxonomy, by = "orf") %>%
    arrange(desc(logFC))
  ) %>%
   kable_styling(bootstrap_options = c("striped", "hover")) %>%
    add_header_above(c("Upregulated genelist orfs in TN-treatment" = 22), font_size = 18) %>%
  scroll_box(width = "100%", height = "500px")
```
The taxonomy of genes match the increase in taxonomy on the order level seen in the TN-treatment as opposed to the C-treatment (Burkholderiales and somewhat Flavobacteriales).

```{r, message=F, warning=FALSE}
kable(target_genes_present %>%
  filter(type == "Down") %>%
  select(-Category) %>%
    inner_join(taxonomy, by = "orf") %>%
    arrange(desc(logFC))
  ) %>%
   kable_styling(bootstrap_options = c("striped", "hover")) %>%
    add_header_above(c("Upregulated genelist orfs in C-treatment" = 22), font_size = 18) %>%
  scroll_box(width = "100%", height = "500px")
```
Among the Down-regulated genes, oxidative stress and a type 1 restriction modification as well as an osmoregulatory gene (katG) were found to be downregulated in the TN-treatment, ATP-SYNTHASES from alpha-and gamma-proteobacteria were also found to be upregulated.Furtherrmore, there are no GO-terms related for blebbing or OMV's in the up/downregulated genes (GO-terms GO:0032060,GO:0061701). The same was true for antibiotic response term GO:0046677.
```{r, message=F, warning=FALSE}
# Identify genes related to blebbing, look at orfs annotated as GO:0032060 in the up/downregulated genes.
# Genes related to OMV's on the other hand GO:0061701 exist.. Nor was anything related to antibiotic responses found through the GO-term GO:0046677
edgeR_orfs %>%
  #filter(type %in% c("Up","Down")) %>%
  inner_join(
    eggnogs %>% distinct(orf, GOs),
    by = "orf"
  ) %>%
  filter(GOs %in% c("GO:0032060","GO:0061701","GO:0046677"))
```
## General orf-response
```{r, message=F, warning=FALSE, fig.width=8}
#Looking at functions of the orfs and also doing this by taxonomy later
edgeR_orfs %>%
  filter(type %in% c("Up","Down")) %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(class, type) %>%
  summarise(SumlogCPM = sum(logCPM)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x = class, y = SumlogCPM)) +
  geom_col() +
  facet_wrap(~ type) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  ggtitle("summarised logCPM of class for up and down regulated genes")
```

### Upregulated genes
Out of the upregulated genes, only `r edgeR_orfs %>% filter(type == "Up") %>% inner_join(eggnogs %>% distinct(orf,Preferred_name),by = "orf") %>% filter(Preferred_name != "-") %>% nrow()` are annotated.
Here follows some example genes of interest, for personal perusing, consort the table below. These genes include impA (a type VI secretion protein)    
pstS (many pstS genes mainly in alpha proteobacteria, one or two betas also have this gene upregulated),phoB, part of a phosphate high affinity system and starvation protein complex respectively.   
phnCD phosphonate uptake system   
amtB ammonium transporter   
[pufB](https://www.uniprot.org/uniprotkb/P0C0Y1/entry) a light harvesting protein.      
[flgE](https://www.uniprot.org/uniprotkb/P75937/entry) a flagellar hook protein (several copies).
[flgC](https://www.uniprot.org/uniprotkb/P0ABX2/entry) motility.  
[fliC](https://www.uniprot.org/uniprotkb/P04949/entry) flagellin gene, motility.However, in some bacteria flagellin is not necessarily used for mtility, but [may be needed for autoaggreagation](https://www.frontiersin.org/articles/10.3389/fmicb.2017.00301/full), which perhaps can explain the expression of this gene and the aggregations and connections seen in the Electron microscopy...  
[fliD](https://www.uniprot.org/uniprotkb/P24216/entry) motility.
 virD4 considered a [T4SS](https://en.wikipedia.org/wiki/Type_IV_secretion_system) protein, matches the cell interaction potentially. 
 While not included in the target gene list ftsZ is upregulated in the TN-treatment.   
impA [T6SS](https://en.wikipedia.org/wiki/Type_VI_secretion_system)
[obg](https://www.uniprot.org/uniprotkb/P20964/entry) Involved in  sporulation, is it possible that sporulation can produce different morphologies??

I further checked the presence of antibiotics production ,as this has been shown to cause aggregation in aquatic communities [Ref](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4077313/), but a quick look yielded nothing. On the other hand one or two resistance genes were found in both treatments...
The table is ordered by logFC.
```{r, message=F, warning=FALSE}
kable(
  edgeR_orfs %>%
  filter(type == "Up") %>%
  inner_join(
    eggnogs %>% distinct(orf, Preferred_name, best_og_desc),
    by = "orf") %>%
  filter(Preferred_name != "-") %>% #In total 418 are annotated
  mutate(Preferred_name = forcats::as_factor(Preferred_name) %>% fct_reorder(logFC, .desc = TRUE)) %>%
    inner_join(taxonomy, by = "orf") %>%
    arrange(desc(logFC))
  ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  add_header_above(c("Annotated upreguated orfs with gene name and taxonomy" = 22), font_size = 18) %>%
  scroll_box(width = "100%", height = "500px")
```
### Downregulated genes
Out of the `r edgeR_orfs %>% filter(type == "Down") %>% nrow()`downregulated genes, only `r edgeR_orfs %>% filter(type == "Down") %>% inner_join(eggnogs %>% distinct(orf,Preferred_name),by = "orf") %>% filter(Preferred_name != "-") %>% nrow()` are annotated. The taxonomic partitioning of Differentially expressed genes in the C-treatment was more diverse (see table below), with taxa including `r edgeR_orfs %>% filter(type == "Down") %>% inner_join(eggnogs %>% distinct(orf, Preferred_name, best_og_desc), by = "orf") %>% filter(Preferred_name != "-") %>% mutate(Preferred_name = forcats::as_factor(Preferred_name) %>% fct_reorder(logFC, .desc = TRUE)) %>% inner_join(taxonomy, by = "orf") %>% distinct(class) %>% pull(class) ` There are in this list many genes related to the ribosome.   
The fliC gene is also upregulated in this treatment, but fewer. For full list see the table below. Genes are once again ordered by logFC.   
The [srmB](https://www.uniprot.org/uniprotkb/P21507/entry) gene is required for ribosome assembly at low temperatures.
```{r, message=F, warning=FALSE}
kable(
  edgeR_orfs %>%
  filter(type == "Down") %>%
  inner_join(
    eggnogs %>% distinct(orf, Preferred_name, best_og_desc),
    by = "orf") %>%
  filter(Preferred_name != "-") %>% #In total 418 are annotated
  mutate(Preferred_name = forcats::as_factor(Preferred_name) %>% fct_reorder(logFC, .desc = TRUE)) %>%
    inner_join(taxonomy, by = "orf") %>%
    arrange(desc(logFC))
  ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  add_header_above(c("Annotated upreguated orfs with gene name and taxonomy" = 22), font_size = 18) %>%
  scroll_box(width = "100%", height = "500px")
```
### Archaea (personal interest)
In this section I am focusing a bit on the archaeal response, as we have such a large fraction of archaea in the C treatment as opposed to the TN treatment.
Out of the `r edgeR_orfs %>% filter(type == "Down") %>% inner_join(taxonomy, by = "orf") %>% filter(domain == "Archaea") %>% nrow()` archaeal genes, only `r edgeR_orfs %>% filter(type == "Down") %>% inner_join(taxonomy, by = "orf") %>% filter(domain == "Archaea") %>% inner_join(eggnogs %>% distinct(orf, Preferred_name, best_og_desc),by = "orf") %>% filter(Preferred_name != "-") %>% nrow()` were annotated. Potentially we are missing important genes here. Interestingly, archaeal pili genes are [potentially very different](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2570433/pdf/nihms62201.pdf) from bacterial, and may not be properly identified with the eggnog-mapper if they are not well cahracterized. Indeed the flagellin in archaea ha similarities to the type IV pilus system, and homology search is not enough to identify them, suggesting novel mechanisms [Ref.](https://pubmed.ncbi.nlm.nih.gov/16983194/)
```{r, message=F, warning=FALSE}
kable(
  edgeR_orfs %>%
  filter(type == "Down") %>%
  inner_join(
    eggnogs %>% distinct(orf, Preferred_name, best_og_desc),
    by = "orf") %>%
  filter(Preferred_name != "-") %>% #In total 418 are annotated
  mutate(Preferred_name = forcats::as_factor(Preferred_name) %>% fct_reorder(logFC, .desc = TRUE)) %>%
    inner_join(taxonomy, by = "orf") %>%
    filter(domain == "Archaea")
  ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  add_header_above(c("Annotated upreguated orfs with gene name and taxonomy" = 22), font_size = 18) %>%
  scroll_box(width = "100%", height = "500px")
```
# Summary of interesting patterns from the gene expression analysis
Excerpt from the powerpoint presented by Ashish an Johan:   
"Blebbing of outer membrane, and release of membrane like vesicles are two characters which were significantly higher in TN treatment as compared to C, while the presence of pili all over surface were significantly higher in C in contrast to TN treatment.
Connections between cells in combination with blebbing of outer membrane(Con + Blb) was the only combination which were significantly higher in TN in contrast to control."   

Supporting these patterns:
Among the genes not in the target list, but related to the morphological differences, upregulation of fliC may be an indication of cell connection (aggregation). Two genes related to difrferent secretion systems were also noted in the TN-treatment (impA, virD4), potentially related to the cell connections. No genes related to conjugation (tra-genes) were upregulated in the TN-treatment. Furthermore, one ccmA gene was upregulated in the TN treatment (related to cell shape).   

On the note of processes related to maintenance metabolism, the downregulated genes (upregulated in C-treatment) included the osmoregulatory gene proW, a type 1 restriction-modification gene hsdM, and two copies of the katG gene involved in oxidative stress. WHile the ATP genes were seen in both up and downregulated genes, but expressed by different taxa, suggesting a potential indicator for growth. The pepA was uprregulated in the C-treatment, which is related to intracellular protein turnover.


