---
title: "EdgeR"
author: "Dennis Amnebrink"
date: '2022-09-06'
output:
  html_document:
      toc: yes
      toc_float:
        collapse: no
      fig_caption: yes
      code_folding: hide
---

EdgeR-analysis of contrast TNvsC for days 10 and 17 respectively. The analysis is carried out on KO-terms, which is a functional annotation based on the [KEGG-system.](https://www.genome.jp/kegg/)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(data.table)
library(edgeR)
library(kableExtra)
```

```{r file-loading, message=FALSE, warning=FALSE, cache=TRUE}
#File to connect NGI ID with our own classification ID, add the Âµ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz")

#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm) 
```

```{r constants, message=FALSE, warning=FALSE}
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r prok-filter, message=FALSE, warning=FALSE, cache=TRUE}
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```

## EdgeR-analysis day 10 
```{r edgeR_day10, message=FALSE, warning=FALSE}
# Do T vs C only day 10 (T3)
 bbmap_mat <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>% #Keeping day 10
  select(sample, KEGG_ko, count) %>%
  spread(sample, count, fill = 0) %>%
  column_to_rownames("KEGG_ko") %>%
  as.matrix()
 

y <- DGEList(bbmap_mat)

#Changing the grouping to the known treatments from mesocosm.
y$samples$group = sample_ID %>% filter(day == "10") %>% select(sample, treatment) %>%  pull(treatment)

#It has been correctly done as they match with the original sample file
groupF <- y$samples$group

#Another way of verifying
match(rownames(y$samples), colnames(bbmap_mat))

y <- DGEList(counts=bbmap_mat, group=groupF)

keep <- filterByExpr(y, group=groupF)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y$samples

#The design-matrix doesn't like non-numeric categories, so here is a numeric corresponding to orignal levels, key =
# 1 = T, 2 = C, 3 = N, 4 = TN
design_m <- y$samples %>% select(group) %>%
  mutate(group = c(1,1,1,2,2,2,3,3,3,4,4,4))

y <- estimateDisp(y, design = design_m)

plotBCV(y)

#Creating a designmatrix (this looks like the correct structure), C is the base treatment, excluding intercept column.
trial_mat <- model.matrix(~0+groupF)

fit <- glmQLFit(y, trial_mat)

fit

#Comparing TN vs C
colnames(trial_mat) = c("C","N","TN","t")

fit <- glmQLFit(y, trial_mat)

TNvsC <- makeContrasts(TN - C, levels=trial_mat)
```

```{r, adding_glm, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TNvsC)

day10 <- topTags(qlf, n = NULL) %>%
  as.data.frame() %>%
  mutate(day = "10") %>%
  rownames_to_column("KEGG_ko") %>%
  select(-FDR)
```

## EdgeR-analysis day 17
```{r, edgeR_day17, message=FALSE, warning=FALSE}
 bbmap_mat <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  select(sample, KEGG_ko, count) %>%
  spread(sample, count, fill = 0) %>%
  column_to_rownames("KEGG_ko") %>%
  as.matrix()

y <- DGEList(bbmap_mat)

#Changing the grouping to the known treatments from mesocosm.
y$samples$group = sample_ID %>% filter(day == "17") %>% select(sample, treatment) %>%  pull(treatment)

#It has been correctly done as they match with the original sample file
groupF <- y$samples$group

#Another way of verifying
match(rownames(y$samples), colnames(bbmap_mat))

y <- DGEList(counts=bbmap_mat, group=groupF)

keep <- filterByExpr(y, group=groupF)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y$samples

#The design-matrix doesn't like non-numeric categories, so here is a numeric corresponding to orignal levels, key =
# 1 = T, 2 = C, 3 = N, 4 = TN
design_m <- y$samples %>% select(group) %>%
  mutate(group = c(1,1,1,2,2,2,3,3,3,4,4,4))

y <- estimateDisp(y, design = design_m)

plotBCV(y)

#Creating a designmatrix (this looks like the correct structure), C is the base treatment, exlcuding intercept column.
trial_mat <- model.matrix(~0+groupF)

fit <- glmQLFit(y, trial_mat)

fit

#Comparing C vs T
colnames(trial_mat) = c("C","N","TN","t")

fit <- glmQLFit(y, trial_mat)

TNvsC <- makeContrasts(TN - C, levels=trial_mat)
```

```{r, joining_contrasting_days, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TNvsC)

day17 <- topTags(qlf, n = NULL) %>%
  as.data.frame() %>%
  mutate(day = "17") %>%
  rownames_to_column("KEGG_ko") %>%
  select(-FDR)

 edgeR_list <- day10 %>%
  bind_rows(day17) %>%
  mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down"))
```

```{r summary_edgeR_tests}
kable(edgeR_list %>%
  group_by(day, type) %>%
  tally() %>%
  ungroup() %>%
  spread(day, n, fill = 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
    add_header_above(c("Table 1. EdgeR stats in contrast TNvsC for day 10 and 17 respectively" = 3), font_size = 18)
```

```{r edgeR_plot, message=F, warning=FALSE, fig.cap="Gene abundance against logFC, Up denotes upregulated in the TN treatment, down indicates upregulated in the C-treatment. Horizontal dashed green lines indicates `r DIFF_TRESHOLD` logFC treshold."}
#Setting colours
colours <- c("#ff0000","#000000","#4248f5")

edgeR_list %>%
  ggplot(mapping = aes(x = logFC, y = logCPM)) +
  geom_point(aes(colour = type, size = type)) +
  scale_color_manual("Type",values = colours) +
  scale_size_manual(values = c("Up" = 1.5, "non-DE" = 0.2, "Down" = 1.5), guide = "none") +
  geom_vline(xintercept = DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  geom_vline(xintercept = -DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  theme_classic() +
  coord_flip() +
  facet_wrap(~ day)
```
```{r upregulated_genes, fig.cap="logCPM distribution of the genes per day and coloured by treatment."}
edgeR_list %>%
  filter(type %in% c("Up","Down")) %>% 
  ggplot(mapping = aes(x = day, y = logCPM,  fill = type)) +
  geom_boxplot() +
  theme_minimal()
```

<br>

No overlaps were seen between up and downregulated KO's
```{r overlaps_Down_Up}
#Matching day 10
match(edgeR_list %>%
  filter(type %in% c("Up")) %>%
  filter(day == "10") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko),

edgeR_list %>%
  filter(type %in% c("Down")) %>%
  filter(day == "10") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko)
)
# Matching day 17
match(edgeR_list %>%
  filter(type %in% c("Up")) %>%
  filter(day == "17") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko),

edgeR_list %>%
  filter(type %in% c("Down")) %>%
  filter(day == "17") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko)
)

```

The KO-gene lists created were uploaded to the [KEGG-pathway](https://www.genome.jp/kegg/mapper/reconstruct.html) reconstruction for quick insight into genes upregulated. Briefly the day 10 genes upregulated in the TN-treatment included five genes related to N-metabolism, such that complete denitrification (NO3 -> N2) was possible. Additionally four ABC-transporters were upregulated, belonging to eukaryotic ABC-transporters, which doesn't make sense as all non-prokaryotic reads have been removed.

```{r Overlaps_Up}
up10 <- edgeR_list %>%
  filter(type %in% c("Up")) %>%
  filter(day == "10") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko)

up17 <- edgeR_list %>%
  filter(type %in% c("Up")) %>%
  filter(day == "17") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  pull(KEGG_ko)

matches <- up10[!is.na(up17[match(up10, up17)])]

```
Out of the Upregulated genes in the TN-treatment, only `r length(matches)` were shared, these were `r matches`.


```{r gene_kegg_table_Up}
edgeR_list %>%
  filter(type == "Up") %>%
  filter(day == "10") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/day10_Up.tsv")

edgeR_list %>%
  filter(type == "Up") %>%
  filter(day == "17") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/day17_Up.tsv")
```

```{r gene_kegg_table_Down}
edgeR_list %>%
  filter(type == "Down") %>%
  filter(day == "10") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/day10_Down.tsv")

edgeR_list %>%
  filter(type == "Down") %>%
  filter(day == "17") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/day17_Down.tsv")
```

