---
title: "edgerPref"
author: "Dennis Amnebrink"
affiliation-title: "Linnaeus University"
title-block-banner: true
date: now
editor: visual
toc: true
prefer-html: true
number-sections: true
format:
  html:
     code-fold: true
     self-contained: true
---

# EdgeR-analysis

EdgeR-analysis of contrast TNvsC for day 17. The analysis is carried out on orf-level, allowing for identification of taxa specific responses.

```{r}
#| warning: false
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
MIN_COUNT = 15
MIN_SAMPLE = 2
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r}
#| warning: false
library(tidyverse)
library(data.table)
library(edgeR)
library(kableExtra)
library(pals)
library(viridis)
library(patchwork)
```

```{r}
#| warning: false
#| cache: true
#File to connect NGI ID with our own classification ID, add the Âµ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#COG to gene-name file
cognames <- read_tsv("../data/cognames.tsv", col_names = c("eggNOG OGs","cogname"))
#Target gene file
target_genes <- read_tsv("../data/list_of_genes.tsv")

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz") %>% # Overwriting the Preferred_names with COGs of interest, for a priori list...
  mutate(`eggNOG OGs` = gsub("@...*","",`eggNOG OGs`)) %>%
  mutate(Preferred_name = case_when(
      `eggNOG OGs` %in% target_genes$gene[grepl("COG", target_genes$gene)] ~ `eggNOG OGs`,
      TRUE ~ Preferred_name
  ))


#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm)

rm_values <- read_tsv("../data/Rm_values.tsv")


gene_categories <- read_tsv("../data/gene_categories.tsv")
```

```{r}
#| warning: false
#| cache: true
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```

# EdgeR - analysis of genes summarised by "preferred_name" 
```{r edgeR_2, message=F, warning=FALSE}
design <- model.matrix(~ 0 + factor(c(1,1,1,2,2,2,3,3,3,4,4,4)))
colnames(design) <- c("warm_control","cold_control","cold_nutrient","warm_nutrient")

#Calculating library size
lib_sizes <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  group_by(orf) %>%
  filter(sum(count) >= MIN_COUNT, n() >= MIN_SAMPLE) %>% #Removing low abundant orfs from analysis
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  inner_join(eggnogs, by = "orf") %>%
  group_by(sample, Preferred_name) %>%
  summarise(counts = sum(count)) %>%
  ungroup() %>%
  select(sample, Preferred_name, counts) %>%
  pivot_wider(names_from = sample, values_from = counts, values_fill = 0) %>%
  column_to_rownames("Preferred_name") 

#Creating DGElist
dgelist <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  group_by(orf) %>%
  filter(sum(count) >= MIN_COUNT, n() >= MIN_SAMPLE) %>% #Removing low abundant orfs from analysis
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  inner_join(eggnogs, by = "orf") %>%
  group_by(sample, Preferred_name) %>%
  summarise(counts = sum(count)) %>%
  ungroup() %>%
  select(sample, Preferred_name, counts) %>%
  pivot_wider(names_from = sample, values_from = counts, values_fill = 0) %>%
  filter(Preferred_name != "-") %>%
  column_to_rownames("Preferred_name") %>%
  DGEList(lib.size = colSums(lib_sizes)) %>%
  calcNormFactors() %>%
  estimateDisp(design) %>%
  glmQLFit(design, robust = TRUE)

#Creating contrasts
my.contrasts <- makeContrasts(warm_nutrientvscold_control=warm_nutrient-cold_control, levels=design)

qlf.TNvsC <- glmQLFTest(dgelist, contrast=my.contrasts[,"warm_nutrientvscold_control"])
```

```{r edgeR_2_tsv, message=F, warning=FALSE}
topTags(qlf.TNvsC, n = Inf) %>%
as.data.frame() %>%
  mutate(day = "17") %>%
  rownames_to_column("Preferred_name") %>%
  select(-FDR) %>%
  mutate(contrast = "TNvsC") %>%
  mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down")) %>%
  write_tsv("../results/edgeR_preferred_name.tsv")
```

## Sanity check
```{r edgeR_2_sanity_check, message=F, warning=FALSE}
edgerPrefName <- read_tsv("../results/edgeR_preferred_name.tsv")

upp <- edgerPrefName %>%
  filter(type == "Up") %>%
  arrange(desc(logFC)) %>%
  slice(1:20) %>%
  pull(Preferred_name)

ner <- edgerPrefName %>%
  filter(type == "Down") %>%
  arrange(desc(logFC)) %>%
  slice(1:20) %>%
  pull(Preferred_name)

bbmap_p %>% 
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  inner_join(eggnogs, by = "orf") %>%
  group_by(sample, Preferred_name) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(treatment %in% c("TN","C")) %>%
  semi_join(edgerPrefName, by = "Preferred_name") %>%
  group_by(Preferred_name, treatment) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(edgerPrefName, by = "Preferred_name") %>%
  filter(Preferred_name %in% c(target_genes$gene, upp, ner)) %>%
  ggplot(mapping = aes(x = Preferred_name, y = tpm, fill = treatment)) +
  geom_col() + 
  facet_wrap(~ type) +
  theme(axis.text.x = element_text(hjust = 1, angle = 60)) +
  ylim(0, 5000)
```
        
Table showing the number of downregulated, upregulated and non-differentially expressed.
```{r overview_stats_list_of_genes, message=F, warning=FALSE}
kable(edgerPrefName %>%
  group_by(type) %>%
  tally() %>%
  ungroup()
) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```
Volcano plot of genes from DE-analysis
```{r volcano_plot, message=F, warning=FALSE, fig.cap="Volcano plot of all genenames included in the analysis. X-axis is logCPM (abundance), Y-axis denote log-fold change. Dotted lines mark the minimum logFC of 2.5"}
colours <- c("#ff0000","#000000","#4248f5")

edgerPrefName %>%
  ggplot(mapping = aes(x = logFC, y = logCPM)) +
  geom_point(aes(colour = type, size = type)) +
  scale_color_manual("Type",values = colours) +
  scale_size_manual(values = c("Up" = 1.5, "non-DE" = 0.2, "Down" = 1.5), guide = "none") +
  geom_vline(xintercept = DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  geom_vline(xintercept = -DIFF_TRESHOLD, colour = "green", linetype = "dashed") +
  theme_classic() +
  coord_flip() 
```

Number of target genes based on a-priori gene list worked out by Ashish & Dennis in each category, instead of all genes involved in the analysis. Notably, no genes are downregulated (more expressed in the C-treatment). The issue with this analysis is the lack of genes in the target-list, nly nine genes are upregulated, and none are downregulated.
```{r list_of_genes, message=F, warning=FALSE}
kable(edgerPrefName %>%
  rename(gene = "Preferred_name") %>%
  inner_join(target_genes, by ="gene") %>%
  group_by(type) %>%
  tally() %>%
  ungroup() 
) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

Here the same list as above but with gene names included.   
```{r gene_list_DE_genes, message=F, warning=FALSE}
DT::datatable(
  edgerPrefName %>%
  rename(gene = "Preferred_name") %>%
  inner_join(target_genes, by ="gene") %>%
  arrange(desc(type))
  ) %>%
   DT::formatStyle(columns = colnames(df), fontSize = '150%')
```
The majority of genes from the list were not differentially expressed between treatments. A total of 9 genes were upregulated in the TN-treatment, these were involved in Lipid transport and metabolism, restriction modification (type 1 and 3), crispr-cas, oxidative stress, outer membrane vesicles and lastly, mRNA syntehsis and lastly Oxidative stress.

## Plotting the most abundant upregulated genes
Here the top50 most abundant and upregulated genes are plotted and investigated.
```{r, message=F, warning=FALSE}

upp <- edgerPrefName %>%
  filter(type == "Up") %>%
  arrange(desc(logCPM)) %>%
  slice(1:50) %>%
  pull(Preferred_name)

bbmap_p %>% 
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "17") %>% #Keeping day 17
  inner_join(eggnogs, by = "orf") %>%
  group_by(sample, Preferred_name) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(treatment %in% c("TN","C")) %>%
  semi_join(edgerPrefName, by = "Preferred_name") %>%
  group_by(Preferred_name, treatment) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(edgerPrefName, by = "Preferred_name") %>%
  filter(Preferred_name %in% upp) %>%
  mutate(Preferred_name = as.factor(Preferred_name) %>% fct_reorder(logCPM, .desc = TRUE)) %>%
  ggplot(mapping = aes(x = Preferred_name, y = tpm, fill = treatment)) +
  geom_col() + 
  facet_wrap(~ type) +
  theme(axis.text.x = element_text(hjust = 1, angle = 60))
```

## How large is the fraction of the transcriptome that the genes are responsible for in the respective treatments??
```{r}
bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(treatment %in% c("C","TN")) %>%
  group_by(orf, treatment) %>%
  summarise(mean_tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(eggnogs %>% distinct(orf, Preferred_name), by = "orf") %>%
  filter(Preferred_name %in% (edgerPrefName %>% filter(type == "Up") %>% pull(Preferred_name))) %>%
  group_by(treatment) %>%
  summarise(sum_tpm = sum(mean_tpm)) %>%
  ungroup() 

## In the TN-treatment the Upregulated genes were on average responsible for 19514.315 tpm
 bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(treatment %in% c("C","TN")) %>%
  group_by(orf, treatment) %>%
  summarise(mean_tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(eggnogs %>% distinct(orf, Preferred_name), by = "orf") %>%
  filter(Preferred_name %in% (edgerPrefName %>% filter(type == "Down") %>% pull(Preferred_name))) %>%
  group_by(treatment) %>%
  summarise(sum_tpm = sum(mean_tpm)) %>%
  ungroup() 
```

### Creating supplementary table
```{r}
#This table contains logFC, logCPM, up or down regulation, add the category info
edgerPrefName %>%
  left_join(gene_categories, by = "Preferred_name") %>%
  write_tsv("../results/DE_supp_table.tsv")
```


After shortly going through the list of the top50, these few genes I deemed potentially relevant in relation to morphological differences,  due to the abundance perspective;       
   
   
[cpsB](https://www.uniprot.org/uniprotkb/Q54518/entry) - capsule polysaccharide biosynthetic process   
[skp](https://www.uniprot.org/uniprotkb/P0AEU7/entry) - chaperonin required for efficient release of ompA from inner membrane.   
[tonB2](https://www.uniprot.org/uniprotkb/Q9RMT3/entry) - Interacts with outer membrane receptor proteins, uptake into periplasm.   
[yfmR](https://www.uniprot.org/uniprotkb/O06476/entry) - ABC transporter.   
```{r eval=FALSE}
eggnogs %>% 
  distinct(Preferred_name, best_og_desc) %>%
  filter(Preferred_name %in% upp) %>%
  select(Preferred_name, best_og_desc) %>%
  view("top50_upp")
```

Other interesting metabolism genes among the top50:   
[fadEHIN](https://www.uniprot.org/uniprotkb/Q47146/entry) - fatty acid beta oxidation     

## The rest of the upregulated genes
Here I filtered the genelist on upregulated genes and went through the genenames with the statistical difference in mind, and tried to find genes that could support what is found in the morhplogy, mainly blebbing and cell-cell connections.   
Excerpt from the powerpoint presented by Ashish an Johan:   
"Blebbing of outer membrane, and release of membrane like vesicles are two characters which were significantly higher in TN treatment as compared to C, while the presence of pili all over surface were significantly higher in C in contrast to TN treatment.
Connections between cells in combination with blebbing of outer membrane(Con + Blb) was the only combination which were significantly higher in TN in contrast to control."   
## All genenames from DE-analysis (Upregulation)
```{r DE_genes_overview, message=F, warning=FALSE}
DT::datatable(
  edgerPrefName %>%
  rename(gene = "Preferred_name") %>%
  arrange(desc(type)) %>%
    select(-day, -contrast, -PValue),
  filter = list(
    position = 'top', clear = FALSE
    )
  ) %>%
  DT::formatStyle(columns = colnames(df), fontSize = '150%')
```
Focusing on the whole list, the upregulated genes also contained:
[fadI](https://www.uniprot.org/uniprotkb/Q7N287/entry) - involved in FA-oxidation.   
[amsD](https://www.uniprot.org/uniprotkb/Q46634/entry) - Exopolysaccharide biosynthesis   
[arcA](https://www.uniprot.org/uniprotkb/O96622/entry) - part of Arp2/3 complex related to cell motility and chemotaxis.   
[casC](https://www.uniprot.org/uniprotkb/Q46899/entry) - CRISPR-related.   
[cheB](https://www.uniprot.org/uniprotkb/P07330/entry) - chemoxatis/motility.   
[cheC](https://www.uniprot.org/uniprotkb/P40403/entry) - chemotaxis   
[cheV](https://www.uniprot.org/uniprotkb/P37599/entry) - chemotaxis   
[cheX](https://www.uniprot.org/uniprotkb/Q9X1V3/entry) - chemotaxis   
[cheY](https://www.uniprot.org/uniprotkb/P0AE67/entry) - chemotaxis   
[cheZ](https://www.uniprot.org/uniprotkb/P0A9H9/entry) - chemotaxis and motility   
[chvE](https://www.uniprot.org/uniprotkb/P54082/entry) - chemotaxis   
[cmr6](https://www.uniprot.org/uniprotkb/Q8U1T1/entry) - CRISPR   
[cpsB](https://www.uniprot.org/uniprotkb/Q54518/entry) - capsule polysaccharide biosynthetic process   
[cpsE](https://www.uniprot.org/uniprotkb/Q04664/entry) - capsule polysaccharide biosynthetic process   
[ctrA](https://www.uniprot.org/uniprotkb/P0A0V9/entry) - involved in transport of capsular polysaccharides to cell surface.    
"Capsular polysaccharides may promote adherence of bacteria to both surfaces and other bacterial cells, which may facilitate colonization of a particular niche and may lead to the formation of biofilms [13]. Cell-surface polysaccharides have been shown to mediate the attachment of bacterial cells to one another, leading to biofilm formation and persistence of the organisms during colonization [1, 14]." - https://www.intechopen.com/chapters/40582   
[flaA](https://www.uniprot.org/uniprotkb/P27053/entry) - Flagellin   
[fleS](https://www.uniprot.org/uniprotkb/Q9I4N4/entry) - flagellar synthesis, adhesion, swarming motility and antibiotic resistance.   
[flgN](https://www.uniprot.org/uniprotkb/P0A1J7/entry) - flagellum   
[hdc](https://www.uniprot.org/uniprotkb/Q56581/entry) - siderophore THIS IS INTERESTING FORM MORPHOLOGY   
[hipA](https://www.uniprot.org/uniprotkb/P23874/entry) - TA-system, toxin/antitoxin, dormancy process    
[hmuR](https://www.uniprot.org/uniprotkb/Q56989/entry) - Siderophore related   
[hsdR](https://www.uniprot.org/uniprotkb/P08956/entry) - type1 restriction enzyme subunit   
[hsdS](https://www.uniprot.org/uniprotkb/P05719/entry) - type1 restriction system as well   
[kdkA](https://www.uniprot.org/uniprotkb/Q7MPR5/entry) - LPS biosynthesis   
[kmo](https://www.uniprot.org/uniprotkb/Q84HF5/entry) - siderophore synthesis related   
[kpsS](https://www.uniprot.org/uniprotkb/P42218/entry) - capsule polysaccharide export protein.
[lpxD1](https://www.uniprot.org/uniprotkb/Q5X0C0/entry) - involved in synthesis of lipid A, glycolipid anchoring LPS to outer membrane of cell   
[lpxL](https://www.uniprot.org/uniprotkb/P0ACV0/entry) - LPS synthesis   
[lytG](https://www.uniprot.org/uniprotkb/O32083/entry) - Role in motility, chemotaxis and cell division   
[manA](https://www.uniprot.org/uniprotkb/P00946/entry) - capsular polysaccharide related   
[motX](https://www.uniprot.org/uniprotkb/P40608/entry) - Sodium-type polar flagellar protein motX   
[motY](https://www.uniprot.org/uniprotkb/Q9S3P9/entry) - Sodium-type polar flagellar protein motY   
[mscL](https://www.uniprot.org/uniprotkb/P0A742/entry) - large mechanosensitive channel   
[nosF](https://www.uniprot.org/uniprotkb/P19844/entry) - potentially related to ABC-transport   
[pbp](https://www.uniprot.org/uniprotkb/P39844/entry) - peptidoglycan synthesis among others...   
[pilN](https://www.uniprot.org/uniprotkb/G3XD30/entry) - T4SS
[pilO](https://www.uniprot.org/uniprotkb/G3XD51/entry) - T4SS
[pilP](https://www.uniprot.org/uniprotkb/G3XCX7/entry) - T4SS 
[pilU](https://www.uniprot.org/uniprotkb/G3XCX3/entry) - T4SS 
[raxST](https://www.uniprot.org/uniprotkb/P08764/entry) - Restriction system   
[rfaJ](https://www.uniprot.org/uniprotkb/P27129/entry) - LPS biosynthesis   
[supH](https://www.uniprot.org/uniprotkb/Q8A1G1/entry) - transport of oligosaccharides from surface of Outer Membrane to the periplasm      
[tonB2](https://www.uniprot.org/uniprotkb/Q9RMT3/entry) - transport to inner membrane prot potentially   
[virD4](https://www.uniprot.org/uniprotkb/Q6FYV9/entry) - T4SS   
[waaA](https://www.uniprot.org/uniprotkb/P0AC75/entry) - LPS biosynthesis   
[waaM](https://www.uniprot.org/uniprotkb/P0ACV0/entry) - LPS biosynthesis  
[ybdG](https://www.uniprot.org/uniprotkb/P0AAT4/entry) - Meachanosensitive channel   
[ybis](https://www.uniprot.org/uniprotkb/P0AAX8/entry) - anchoring of Major outermembrane lipoprotein (Braun lipoprotein)   
[ydhB](https://www.uniprot.org/uniprotkb/O05493/entry) - probable membrane transproter protein   
[ykuF](https://www.uniprot.org/uniprotkb/O34717/entry) - FA beta cxidation   
These are some of the genes that I could make sense out of, see the above gene list if you want to look through the genes further.   


Summarising all genenames upregulated in the TN-treatment, the genenames noted may provide support for the patterns seen in the morphology; among others T4SS shows up, chemotaxis may help explain the cell aggregation we see. LPS, siderophore and capsular polysaccharide genes may be supportive in explaining the processes behind blebbing. Is it feasible that chemotaxis is employed for aggregation, or perhaps a response to the aggregation???

## Down-regulated genes


[alaC](https://www.uniprot.org/uniprotkb/P77434/entry) - alanine synthesis   
[bphC1](https://www.uniprot.org/uniprotkb/P47231/entry) - xenobiotic degradation   
[cbbL](https://www.uniprot.org/uniprotkb/O85040/entry) - RuBisCo - large chain   
[cbbQ](https://www.uniprot.org/uniprotkb/Q51858/entry) - RuBisCo related   
[cobI](https://www.uniprot.org/uniprotkb/Q9HZU3/entry) - cobalamin synthesis.   
[cpeA](https://www.uniprot.org/uniprotkb/Q02179/entry) - lightharvesting protein   
[cpeB](https://www.uniprot.org/uniprotkb/Q02180/entry) - lightharvesting protein   
[cpeC](https://www.uniprot.org/uniprotkb/P18542/entry) - related to light harvesting protein   
[ddhA](https://www.uniprot.org/uniprotkb/Q8GPG4/entry) - photoautotrophic growth on DMS as e-donor   
[ddhB](https://www.uniprot.org/uniprotkb/Q8GPG3/entry) - related to anaerobic DMS growth.   
[dnaK2](https://www.uniprot.org/uniprotkb/Q6LS31/entry) - chaperone   
[groL1](https://www.uniprot.org/uniprotkb/P9WPE9/entry) - chaperonin   
[ihfB](https://www.uniprot.org/uniprotkb/P0A6Y1/entry) - viral, conjugation, translational control and more..    
[katA](https://www.uniprot.org/uniprotkb/P29422/entry) - peroxidase   
[lig](https://www.uniprot.org/uniprotkb/P56709/entry) - DNA ligase, recombination and repair   
[marR](https://www.uniprot.org/uniprotkb/P27245/entry) - antibiotic resistance   
[mcpC](https://www.uniprot.org/uniprotkb/P54576/entry) - chemotaxis   
[narJ](https://www.uniprot.org/uniprotkb/P0AF26/entry) - nitrate reductase related chaperone   
[narK](https://www.uniprot.org/uniprotkb/P10903/entry) - nitrite/nitrate uptake   
[nsrR](https://www.uniprot.org/uniprotkb/O07573/entry) - NO-responsive transcriptional regulator   
[phnH](https://www.uniprot.org/uniprotkb/P16686/entry) - phosphonate metabolism   
[porB](https://www.uniprot.org/uniprotkb/P30689/entry) - porin   
[psaA](https://www.uniprot.org/uniprotkb/P56766/entry) - photosystem 1   
[psbA2](https://www.uniprot.org/uniprotkb/P0A446/entry) - photosystem 2   
[rpoK](https://www.uniprot.org/uniprotkb/P39463/entry) - transcription   
[sglT](https://www.uniprot.org/uniprotkb/P96169/entry) - sodium/glucose transporter   
[speE2](https://www.uniprot.org/uniprotkb/Q9X8S2/entry) - spermedine biosynthesis   
[tlpC](https://www.uniprot.org/uniprotkb/P39209/entry) - chemotaxis related   
[wspC](https://www.uniprot.org/uniprotkb/Q88MS8/entry) - probable biofilm formation related   
[ycjP](https://www.uniprot.org/uniprotkb/P77716/entry) - probable ABC transport permease protein   
[ydhH](https://www.uniprot.org/uniprotkb/O07577/entry) - sodium dependent transporter   
[ynjD](https://www.uniprot.org/uniprotkb/P76909/entry) - uncharacterized ABC transporter ATP-binding protein   
[yvcC](https://www.uniprot.org/uniprotkb/O06967/entry) - related to multidrug resistance      
   
The down-regulated genes is a bit harder to summarise as there is a more diverse respoonse. However it appears that genes such as T4SS, LPS biotynthesis, capsular polysaccharides and siderophores are not well represented in comparison to the upregulated genes.   

## Plot of genes with categories

```{r}
p <- edgerPrefName %>%
  inner_join(gene_categories, by = "Preferred_name") %>%
  mutate(Preferred_name = as.factor(Preferred_name) %>% fct_reorder(category)) %>%
  mutate(treatment = ifelse(type == "Up","TN","C")) %>%
  ggplot(mapping = aes(x = Preferred_name, y = logFC, colour = category)) +
  geom_point(aes(size = logCPM)) +
  theme_minimal() +
  theme(axis.text.x = element_text(hjust = 1, angle = 60), legend.position = "bottom", text = element_text(size = 8), legend.spacing.x = unit(0.08, 'cm')) +
  xlab("Genename") +
  scale_colour_manual('Category', values = GEOM_COL_COLOURS) +
  scale_x_discrete(expand = c(0, 0.5)) 

p + ggforce::facet_row(vars(treatment), scales = 'free', space = 'free')

ggsave("../results/diff_genecats.png", bg = "white", height = 8, units = "cm")

```

## How many genes in each category??
```{r}
gene_categories %>%
  group_by(category) %>%
  tally() %>%
  ungroup() %>%
  arrange(desc(n))
```

## Investigate proteorhodopsin related genes
```{r}
# Genes related to proteorhodopsin (from Kimura et al. 2011 ISME)
  #idi -present 
  #ispA - present
  #crtE - present
  #crtB - present
  #crtI - present
  #crtY - present
  #blh - present
# LogCPM around 4 or so, perhaps abundant in all treatments? This would suggest a stable but important function to the system!
```

